pkg/profilestore/profilecolumnstore.go

```diff
@@ -135,8 +141,144 @@ func (s *ProfileColumnStore) writeSeries(ctx context.Context, req *profilestorep
 				return status.Errorf(codes.InvalidArgument, "failed to parse profile: %v", err)
 			}
 
-			if err := ingester.Ingest(ctx, ls, p, req.Normalized); err != nil {
-				return status.Errorf(codes.Internal, "failed to ingest profile: %v", err)
+			if err := ValidatePprofProfile(p); err != nil {
+				return status.Errorf(codes.InvalidArgument, "invalid profile: %v", err)
+			}
+
+			parcacol.LabelNamesFromSamples(
+				ls,
+				p.StringTable,
+				p.Sample,
+				allPprofLabelNames,
+				allPprofNumLabelNames,
+			)
+			samples = append(samples, p)
+		}
+
+		series = append(series, parcacol.Series{
+			Name:    name,
+			Labels:  ls,
+			Samples: samples,
+		})
+	}
+
+	if err := parcacol.NewIngester(
+		s.logger,
+		parcacol.NewNormalizer(s.metastore),
+		s.table,
+		s.schema,
+		s.bufferPool,
+		sortedKeys(allLabelNames),
+		sortedKeys(allPprofLabelNames),
+		sortedKeys(allPprofNumLabelNames),
+	).Ingest(ctx, series, req.Normalized); err != nil {
+		return status.Errorf(codes.Internal, "failed to create ingester: %v", err)
+	}
+
+	return nil
+}
+
+func sortedKeys(m map[string]struct{}) []string {
+	keys := make([]string, 0, len(m))
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	return keys
+}
```

Comment:
```suggestion
func sortedKeys(in map[string]struct{}) []string {
	if len(in) == 0 {
		return []string{}
	}

	out := maps.Keys(in)
	sort.Strings(out)
	return out
}
```
Sadly the sort.Sort isn't generic yet, but when it's going to be, there will only be one of this ever.